From: Kevin Hilman <khilman@mvista.com>
Date: Thu, 21 Sep 2006 23:58:57 +0000 (+0100)
Subject: [ARM] 3856/1: Add clocksource for Intel IXP4xx platforms
X-Git-Url: http://www.kernel.org/git/?p=linux/kernel/git/torvalds/linux-2.6.git;a=commitdiff;h=84904d0ead0a8c419abd45c7b2ac8d76d50a0d48

[ARM] 3856/1: Add clocksource for Intel IXP4xx platforms

Enables the ixp4xx platforms to use Generic time-of-day.

Signed-off-by: Kevin Hilman <khilman@mvista.com>
Acked-by: John Stultz <johnstul@us.ibm.com>
Signed-off-by: Deepak Saxena <dsaxena@plexity.net>
Signed-off-by: Russell King <rmk+kernel@arm.linux.org.uk>
Index: linux-2.6.18/arch/arm/Kconfig
===================================================================
--- linux-2.6.18.orig/arch/arm/Kconfig	2006-10-29 01:59:14.000000000 -0700
+++ linux-2.6.18/arch/arm/Kconfig	2006-10-29 01:59:38.000000000 -0700
@@ -212,6 +212,7 @@
 config ARCH_IXP4XX
 	bool "IXP4xx-based"
 	depends on MMU
+	select GENERIC_TIME
 	help
 	  Support for Intel's IXP4XX (XScale) family of processors.
 
Index: linux-2.6.18/arch/arm/mach-ixp4xx/common.c
===================================================================
--- linux-2.6.18.orig/arch/arm/mach-ixp4xx/common.c	2006-09-19 20:42:06.000000000 -0700
+++ linux-2.6.18/arch/arm/mach-ixp4xx/common.c	2006-10-29 01:00:08.000000000 -0800
@@ -26,6 +26,7 @@
 #include <linux/bitops.h>
 #include <linux/time.h>
 #include <linux/timex.h>
+#include <linux/clocksource.h>
 
 #include <asm/hardware.h>
 #include <asm/uaccess.h>
@@ -255,16 +256,6 @@
 
 #define CLOCK_TICKS_PER_USEC	((CLOCK_TICK_RATE + USEC_PER_SEC/2) / USEC_PER_SEC)
 
-/* IRQs are disabled before entering here from do_gettimeofday() */
-static unsigned long ixp4xx_gettimeoffset(void)
-{
-	u32 elapsed;
-
-	elapsed = *IXP4XX_OSTS - last_jiffy_time;
-
-	return elapsed / CLOCK_TICKS_PER_USEC;
-}
-
 static irqreturn_t ixp4xx_timer_interrupt(int irq, void *dev_id, struct pt_regs *regs)
 {
 	write_seqlock(&xtime_lock);
@@ -309,7 +300,6 @@
 
 struct sys_timer ixp4xx_timer = {
 	.init		= ixp4xx_timer_init,
-	.offset		= ixp4xx_gettimeoffset,
 };
 
 static struct resource ixp46x_i2c_resources[] = {
@@ -365,3 +355,29 @@
 			ixp4xx_exp_bus_size >> 20);
 }
 
+cycle_t ixp4xx_get_cycles(void)
+{
+	return *IXP4XX_OSTS;
+}
+
+static struct clocksource clocksource_ixp4xx = {
+	.name 		= "OSTS",
+	.rating		= 200,
+	.read		= ixp4xx_get_cycles,
+	.mask		= CLOCKSOURCE_MASK(32),
+	.shift 		= 20,
+	.is_continuous 	= 1,
+};
+
+unsigned long ixp4xx_timer_freq = FREQ;
+static int __init ixp4xx_clocksource_init(void)
+{
+	clocksource_ixp4xx.mult =
+		clocksource_hz2mult(ixp4xx_timer_freq,
+				    clocksource_ixp4xx.shift);
+	clocksource_register(&clocksource_ixp4xx);
+
+	return 0;
+}
+
+device_initcall(ixp4xx_clocksource_init);
Index: linux-2.6.18/arch/arm/mach-ixp4xx/nslu2-setup.c
===================================================================
--- linux-2.6.18.orig/arch/arm/mach-ixp4xx/nslu2-setup.c	2006-09-19 20:42:06.000000000 -0700
+++ linux-2.6.18/arch/arm/mach-ixp4xx/nslu2-setup.c	2006-10-29 01:00:08.000000000 -0800
@@ -159,6 +159,8 @@
 
 static void __init nslu2_init(void)
 {
+	ixp4xx_timer_freq = NSLU2_FREQ;
+
 	ixp4xx_sys_init();
 
 	nslu2_flash_resource.start = IXP4XX_EXP_BUS_BASE(0);
Index: linux-2.6.18/include/asm-arm/arch-ixp4xx/platform.h
===================================================================
--- linux-2.6.18.orig/include/asm-arm/arch-ixp4xx/platform.h	2006-09-19 20:42:06.000000000 -0700
+++ linux-2.6.18/include/asm-arm/arch-ixp4xx/platform.h	2006-10-29 01:00:08.000000000 -0800
@@ -90,6 +90,11 @@
 struct sys_timer;
 
 /*
+ * Frequency of clock used for primary clocksource
+ */
+extern unsigned long ixp4xx_timer_freq;
+
+/*
  * Functions used by platform-level setup code
  */
 extern void ixp4xx_map_io(void);
